### PR TITLE
test: add testcontainers

### PR DESCRIPTION
**Pull request by luky116**
### Purpose or design rationale of this PR

use testcontainer instead of start docker engine by  cmd

#436 

### PR title

Your PR title must follow [conventional commits](https://www.conventionalcommits.org/en/v1.0.0/#summary) (as we are doing squash merge for each PR), so it must start with one of the following [types](https://github.com/angular/angular/blob/22b96b9/CONTRIBUTING.md#type):

- [ ] build: Changes that affect the build system or external dependencies (example scopes: yarn, eslint, typescript)
- [ ] ci: Changes to our CI configuration files and scripts (example scopes: vercel, github, cypress)
- [ ] docs: Documentation-only changes
- [ ] feat: A new feature
- [ ] fix: A bug fix
- [ ] perf: A code change that improves performance
- [ ] refactor: A code change that doesn't fix a bug, or add a feature, or improves performance
- [ ] style: Changes that do not affect the meaning of the code (white-space, formatting, missing semi-colons, etc)
- [x] test: Adding missing tests or correcting existing tests

### Deployment tag versioning

Has `tag` in `common/version.go` been updated or have you added `bump-version` label to this PR?

- [x] No, this PR doesn't involve a new deployment, git tag, docker image tag
- [ ] Yes

### Breaking change label

Does this PR have the `breaking-change` label?

- [x] No, this PR is not a breaking change
- [ ] Yes

### CODE REVIEW

File: common/testcontainers/testcontainers.go [hunk 1]
```diff
@@ -0,0 +1,165 @@
...
+)
+
+type TestcontainerApps struct {
+	postgresContainer *postgres.PostgresContainer
+	l1GethContainer   *testcontainers.DockerContainer
+	l2GethContainer   *testcontainers.DockerContainer
+
+	dbClient     *sql.DB
+	DBConfig     *database.DBConfig
+	DBConfigFile string
```
Comment (georgehao) [inline common/testcontainers/testcontainers.go]: I think this is no useful

File: common/testcontainers/testcontainers.go [hunk 2]
```diff
@@ -0,0 +1,165 @@
...
+	"scroll-tech/database"
+	"time"
+
+	"github.com/scroll-tech/go-ethereum/ethclient"
+	"github.com/testcontainers/testcontainers-go"
+	"github.com/testcontainers/testcontainers-go/modules/postgres"
+	"github.com/testcontainers/testcontainers-go/wait"
+)
+
+type TestcontainerApps struct {
```
Comment (georgehao) [inline common/testcontainers/testcontainers.go:16]: add comments

File: common/testcontainers/testcontainers.go (hunk 1 above)
Comment (luky116) [reply common/testcontainers/testcontainers.go]: This is to be compatible with the original logic. I will sort out the fields in this struct.

File: common/testcontainers/testcontainers.go (hunk 2 above)
Comment (luky116) [reply common/testcontainers/testcontainers.go:16]: > add comments

ok

Comment (luky116) [reply common/testcontainers/testcontainers.go:16]: done

File: common/testcontainers/testcontainers.go (hunk 1 above)
Comment (luky116) [reply common/testcontainers/testcontainers.go]: done

File: common/testcontainers/testcontainers.go [hunk 3]
```diff
@@ -0,0 +1,175 @@
...
+	"github.com/testcontainers/testcontainers-go/wait"
+)
+
+// TestcontainerApps testcontainers struct
+type TestcontainerApps struct {
+	postgresContainer *postgres.PostgresContainer
+	l1GethContainer   *testcontainers.DockerContainer
+	l2GethContainer   *testcontainers.DockerContainer
+
+	// common time stamp.
```
Comment (colinlyguo) [inline common/testcontainers/testcontainers.go]:
```suggestion
	// common time stamp in nanoseconds.
```

File: common/testcontainers/testcontainers.go [hunk 4]
```diff
@@ -0,0 +1,175 @@
...
+}
+
+// StartL1GethContainer starts a L1Geth container
+func (t *TestcontainerApps) StartL1GethContainer() error {
+	if t.l1GethContainer != nil && t.l1GethContainer.IsRunning() {
+		return nil
+	}
+	req := testcontainers.ContainerRequest{
+		Image:        "scroll_l1geth",
+		ExposedPorts: []string{"8546/tcp", "8545/tcp"},
```
Comment (colinlyguo) [inline common/testcontainers/testcontainers.go:60]:
```suggestion
		ExposedPorts: []string{"8545/tcp"},
```

Comment (colinlyguo) [reply common/testcontainers/testcontainers.go:60]: I think ws port is not necessary, can try removing it.

File: common/testcontainers/testcontainers.go [hunk 5]
```diff
@@ -0,0 +1,175 @@
...
+}
+
+// StartL2GethContainer starts a L2Geth container
+func (t *TestcontainerApps) StartL2GethContainer() error {
+	if t.l2GethContainer != nil && t.l2GethContainer.IsRunning() {
+		return nil
+	}
+	req := testcontainers.ContainerRequest{
+		Image:        "scroll_l2geth",
+		ExposedPorts: []string{"8546/tcp", "8545/tcp"},
```
Comment (colinlyguo) [inline common/testcontainers/testcontainers.go:84]:
```suggestion
		ExposedPorts: []string{"8545/tcp"},
```

File: common/testcontainers/testcontainers.go [hunk 6]
```diff
@@ -0,0 +1,175 @@
...
+	genericContainerReq := testcontainers.GenericContainerRequest{
+		ContainerRequest: req,
+		Started:          true,
+	}
+	container, err := testcontainers.GenericContainer(context.Background(), genericContainerReq)
+	if err != nil {
+		log.Printf("failed to start scroll_l1geth container: %s", err)
+		return err
+	}
+	t.l1GethContainer, _ = container.(*testcontainers.DockerContainer)
```
Comment (colinlyguo) [inline common/testcontainers/testcontainers.go:73]: `error` or `ok` handling?

File: rollup/cmd/mock_app.go [hunk 7]
```diff
@@ -9,6 +9,7 @@ import (
 
 	"scroll-tech/common/cmd"
 	"scroll-tech/common/docker"
```
Comment (colinlyguo) [inline rollup/cmd/mock_app.go:11]: can remove this import, and also the source files.

File: rollup/cmd/mock_app.go [hunk 8]
```diff
@@ -27,9 +30,8 @@ type MockApp struct {
 	args []string
 }
 
-// NewRollupApp return a new rollupApp manager, name mush be one them.
+// NewRollupApp TODO function will be replaced by NewRollupApp2
 func NewRollupApp(base *docker.App, file string) *MockApp {
```
Comment (colinlyguo) [inline rollup/cmd/mock_app.go:34]: could we keep this and only use `testcontainers`? Is there any blocker here?

**Review by colinlyguo — DISMISSED** [review]
LGTM.

File: common/testcontainers/testcontainers.go (hunk 5 above)
Comment (colinlyguo) [reply common/testcontainers/testcontainers.go:84]: I think ws port is not necessary, can try removing it.

File: common/testcontainers/testcontainers.go [hunk 9]
```diff
@@ -0,0 +1,175 @@
...
+	genericContainerReq := testcontainers.GenericContainerRequest{
+		ContainerRequest: req,
+		Started:          true,
+	}
+	container, err := testcontainers.GenericContainer(context.Background(), genericContainerReq)
+	if err != nil {
+		log.Printf("failed to start scroll_l2geth container: %s", err)
+		return err
+	}
+	t.l2GethContainer, _ = container.(*testcontainers.DockerContainer)
```
Comment (colinlyguo) [inline common/testcontainers/testcontainers.go:96]: same comment. handling `_`.

Comment (luky116) [reply common/testcontainers/testcontainers.go:96]: container must be of testcontainers.DockerContainer type,   so it is okay not to handle _

image

Comment (colinlyguo) [reply common/testcontainers/testcontainers.go:96]: oh, got it.

File: common/testcontainers/testcontainers.go (hunk 6 above)
Comment (luky116) [reply common/testcontainers/testcontainers.go:73]: container must be of *testcontainers.DockerContainer type, so it is okay not to judge _ doke

image

File: rollup/cmd/mock_app.go (hunk 7 above)
Comment (luky116) [reply rollup/cmd/mock_app.go:11]: It is still needed now, and some docker has not been completely replaced. I will submit a few more PRs later.

File: common/testcontainers/testcontainers.go (hunk 9 above)
Comment (luky116) [reply common/testcontainers/testcontainers.go:96]: same as above

File: common/testcontainers/testcontainers.go (hunk 5 above)
Comment (luky116) [reply common/testcontainers/testcontainers.go:84]: In the previous program, port 8545 was used, so it was retained; port 8546 was used for container exploration of testcontainer, so it was retained.

image

image

File: rollup/cmd/mock_app.go (hunk 8 above)
Comment (luky116) [reply rollup/cmd/mock_app.go:34]: It is still needed now, and some docker has not been completely replaced. I will submit a few more PRs later.

File: common/testcontainers/testcontainers.go (hunk 3 above)
Comment (luky116) [reply common/testcontainers/testcontainers.go]: done

File: common/testcontainers/testcontainers.go (hunk 4 above)
Comment (luky116) [reply common/testcontainers/testcontainers.go:60]: the same as below

**Review by colinlyguo — DISMISSED** [review]

**Comment by 0xmountaintop** [comment]
looks good but the tests seem failing occasionally

**Review by georgehao — APPROVED** [review]

**Review by colinlyguo — APPROVED** [review]

**Pull request merged**